From fcbea18c584257adc824cbb1c66cec2f4329869c Mon Sep 17 00:00:00 2001
From: Glenn Washburn <development@efficientek.com>
Date: Mon, 14 Aug 2023 01:03:41 -0500
Subject: [PATCH 15/57] fs/archelp: If path given to grub_archelp_dir() is not
 a directory return error

Specifically, return GRUB_ERR_BAD_FILE_TYPE because this is what is
expected by the ls command when it is given a path to a non-directory.
This fixes a bug where calling ls with a list of non-directory paths
outputs a blank line for each such argument.

Signed-off-by: Glenn Washburn <development@efficientek.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>
---
 grub-core/fs/archelp.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/grub-core/fs/archelp.c b/grub-core/fs/archelp.c
index 998de88b8..c1dcc6285 100644
--- a/grub-core/fs/archelp.c
+++ b/grub-core/fs/archelp.c
@@ -180,6 +180,14 @@ grub_archelp_dir (struct grub_archelp_data *data,
 	  if (p)
 	    *p = 0;
 
+	  if ((*n == 0) && ((mode & GRUB_ARCHELP_ATTR_TYPE)
+			    != GRUB_ARCHELP_ATTR_DIR))
+	    {
+	      grub_error (GRUB_ERR_BAD_FILE_TYPE, N_("not a directory"));
+	      grub_free (name);
+	      goto fail;
+	    }
+
 	  if (((!prev) || (grub_strcmp (prev, name) != 0)) && *n != 0)
 	    {
 	      struct grub_dirhook_info info;
-- 
2.42.0

